### DISCUSSION TITLE
Routing Based On Query Param

### DISCUSSION THREAD

**Discussion by melvinkcx** [fastapi/fastapi#6010] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Query


app = FastAPI()


@app.get("/",
         dependencies=[Depends(dep_a)])
async def root_when_x_is_false(
    x: bool = Query(False)
):
    return {"message": "x is False"}


@app.get("/",
         dependencies=[Depends(dep_b)])
async def root_when_x_is_true(
    x: bool = Query(True)
):
    return {"message": "x is True"}
```

### Description

Hi, I've been looking around on the internet and experimenting with the code, but it seems that what I'm trying to achieve might not be possible. 

Here is what I'm trying to do:

* I have an endpoint `/` that accepts an optional query param `x` of boolean type. 
* I want to define 2 different route handlers and invoke them according to the query param's value.

From the code example, if `/?x=true` is called, the second handler should be invoked. Otherwise, the first handler should be called instead. 

Why can't I merge the route handlers into one? I heard you asked.
Each route handler has its own distinct list of dependencies that do not apply to another handler. In the code example, the handler of `/?x=true` depends on `dep_b` while the other depends on `dep_a`.

I wonder if there is a workaround to achieve my purpose, any suggestions are largely welcome, thank you =)

### Operating System

macOS

### FastAPI Version

0.92.0

### Python Version

3.10

**Comment by williamjamir** [comment, marked as answer]
As far as I know, you cannot have two functions to serve the same resource/endpoint.

One possible workaround for your use case is to create a single endpoint that handles both cases and use a single dependency that handles both `dep_a` and `dep_b` 

Within this dependency, you can check the value of the x parameter and call the appropriate one. 

```python
from fastapi import FastAPI, Depends, Query

app = FastAPI()


async def some_func_a():
    return 1


async def some_func_b():
    return 2


async def dep_c(x: bool = False):
    if x:
        return await some_func_a()
    else:
        return await some_func_b()


@app.get('/')
async def some_router_func(result=Depends(dep_c)):
    return result


if __name__ == '__main__':
    import uvicorn

    uvicorn.run('x:app')
```

I hope this answer your question :)

**Reply by melvinkcx** [reply]
Thanks for the suggestion. I appreciate it =)
